[1-2 단계 방탈출 예약 대기] 띠용(주용은) 미션 제출합니다.

jye326 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
[링크](https://prolog.techcourse.co.kr/studylogs/4396)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
완벽하게 체크하진 못했으나 전반적으로 코드 상 불편한 부분이 없었습니다

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->



### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jye326 committed:
initial commit

- dompoo code 복사!

jye326 committed:
feat: VO를 JPA 칼럼으로 매핑

jye326 committed:
feat: 엔티티를 JPA 테이블로 매핑

jye326 committed:
feat: JPA 리포지토리 구현

jye326 committed:
feat: Reservation 엔티티를 테이블에 매핑, 리포지토리 구현

jye326 committed:
fix: 예약 일자 객체 검증 시점 변경

- 최초 생성 시에만 검증

jye326 committed:
feat: 값 객체 Embeddable 어노테이션 추가

jye326 committed:
test: infrastructure 테스트 jpa 전환

jye326 committed:
fix : 쿼리수정, 어노테이션 추가

jye326 committed:
delete: 미사용 resources 삭제

jye326 committed:
fix: 예약된 시간 응답이 논리적으로 반대였던 오류 해결

jye326 committed:
feat: 내 예약 기능의 html, js 추가

jye326 committed:
feat: 내 예약 api 작성

dompoo committed:
style: 코드 스타일 통일

soominsohn commented on src/main/java/roomescape/infrastructure/JdbcReservationTimeRepository.java:
이제 사용하지 않는 Jdbc관련 코드들은 삭제해보면 어떨까요?
남겨두신 이유가 따로 있나요?

jye326 replied:
변경한 JPA 리포지토리 구현체의 일부 기능이 동작하지 않는 오류 상황이 발생할 경우, 이전의 JDBC 구현체로 문제 부분을 빠르게 대체하여 서비스를 가동할 수 있기 때문에 일종의 백업 용으로 남겨두었습니다😁

soominsohn replied:
네 좋아요. 새로운 기술 스택을 적용할 때 예상치 못한 버그나 동작 오류에 대한 걱정은 충분히 공감됩니다.
다만 우리는 Git을 통해 이전 코드가 모두 버전 관리되고 있기 때문에, 실제로 문제가 생기더라도 되돌릴 수 있는 방법이 있네요.
오히려 사용되지 않는 코드를 코드베이스에 그대로 두면 복잡도가 증가한다는 단점이 있을 것 같아요.

또한 이번 미션의 요구사항은 jdbc 의존성을 spring-data-jpa로 완전히 대체해보는 것이기도 하니, 이 기회에 jdbc 코드를 정리하고 전환해보는 경험을 해보는 건 어떨까요?
```
gradle 의존성 추가
build.gradle 파일을 이용하여 다음 의존성을 대체하세요.
as is: spring-boot-starter-jdbc
to be: spring-boot-starter-data-jpa
```

jye326 replied:
그렇군요! 말씀하신 대로 버전 관리가 되고 있으므로 삭제하였습니다. gradle 의존성도 삭제했습니다!

soominsohn replied:
네~ 반영해주셔서 고맙습니다~

soominsohn commented on src/main/java/roomescape/infrastructure/JpaReservationDao.java:
```diff
@@ -0,0 +1,38 @@
+    @Query("""
+            SELECT DISTINCT r
+              FROM Reservation r
+              JOIN FETCH r.time rt
+              JOIN FETCH r.theme t
+              JOIN FETCH r.user u
```
fetch join을 사용한 이유는 무엇인가요?

jye326 replied:
지연 로딩으로 인한 n+1 문제를 예방하기 위함입니다.`JpaReservationDao`는 엔티티간의 연관 관계 매핑 설정을 모르기 때문에 지연 로딩이 설정되었을 경우를 대비하여 한번의 쿼리 요청으로 모든 엔티티를 불러오도록  fetch join 하였습니다.

soominsohn replied:
네 좋아요!
n+1 문제를 예방하기 위한 다른 방법들에는 무엇이 있을지도 학습해보면 좋을 것 같아요 😄

jye326 replied:
join fetch 외에도 방법이 많군요! 다음 미션 시작 전에 학습해보도록 하겠습니다!

soominsohn commented on src/main/java/roomescape/infrastructure/JpaReservationRepository.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.List;
+import java.util.Optional;
+
+@Primary
```
`@Primary `어노테이션은 어떻게 동작하나요?

jye326 replied:
DI 후보가 여러 개일 때 우선순위를 지정합니다. 현재 `JpaReservationRepository`와 기존 `JdbcReservationRepository `가 모두 `@Repository` 로 지정되어 있는데 이러한 경우 `@Primary` 를 가지는 쪽이 `Bean`으로 등록됩니다. 둘 이상의 충돌 컴포넌트가 `@Primary` 를 가지면 `NoUniqueBeanDefinitionException`이 발생합니다.

soominsohn commented on src/main/java/roomescape/business/model/entity/Reservation.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDate;
+
+@ToString
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@EqualsAndHashCode(of = "id")
+@Getter
+@Entity
```
몇 가지 질문이 있습니다!
- 엔티티 의 기본생성자를 protected로 만들어주셨네요! 엔티티로 지정하려면 기본생성자가 왜 필요할까요?
- `@EqualsAndHashCode(of = "id")` equalsAndHashCode를 재정의한 것으로 보아, 띠용은 id가 같으면 동일한 객체로 판단하신 것 같은데, 어떤 이유인지 궁금합니다.

jye326 replied:
- JPA는 리플랙션을 통해 런타임에 객체를 식별하는데 이를 위해 기본 생성자를 요구합니다. 기존의 public이었던 기본 생성자의 접근 지정자는 protected로 변경하였습니다! 프로젝트 코드에서는 사용하지 말고 JPA를 위한 것임을 드러낼 수 있기 때문입니다!
- 페어의 코드 베이스이고 해당 부분의 정확한 설계 이유를 물어보지 못했습니다만 애플리케이션의 Entity 모델에 RDB의 기본키 식별 개념을 그대로 적용하여 OOP와 RDB의 식별 패러다임 미스매치를 완화하려는 시도로 보입니다.

soominsohn replied:
넵! 잘 알아봐주셨네요~.
왜 JPA는 기본키가 같으면 같은 객체로 판단하는지에 대해서도 학습해보면 좋을 것 같아요. 😄

jye326 replied:
넵! 추가적으로 학습한 내용은 다음과 같습니다. 
JPA, 구체적으로는 영속성 컨텍스트를 관리하는 EntityManger(EM)가 PK를 기준으로 객체를 식별하며, 이는 데이터 베이스가 PK를 기준으로 식별하기 때문입니다. EM은 이러한 DB의 기준을 적용해 객체와 PK를 1:1로 매핑하는 IdentityMap 을 관리합니다. 이를 통해 하나의 PK에는 하나의 인스턴스만 관리하며 객체 참조 일관성을 유지하고, 불필요한 동일 쿼리 요청을 줄입니다.

soominsohn replied:
ㅎㅎ 열심히 학습해주셨군요~ 👍

soominsohn commented on src/main/java/roomescape/business/model/entity/Reservation.java:
```diff
@@ -0,0 +1,50 @@
+@Entity
+public class Reservation {
+
+    @EmbeddedId
+    private final Id id;
```
띠용은 엔티티의 id로 UUID 8자리를 사용하셨는데, 중복이 될 수 있는 가능성은 없을까요?
띠용이 생각하기에 id를 UUID로 생성했을 때의 장점은 무엇일까요?
또, JPA가 지원하는 id 생성방식에는 어떤 것들이 있을까요?

jye326 replied:
1. 16진수 8자리의 랜덤 생성이기 때문에 현재 규모에서는 중복 가능성이 굉장히 희박하다고 생각됩니다. 시스템 규모가 충분히 커진다면 중복 가능성을 고려해야 할 것 같습니다.
2. DB와 무관하게 애플리케이션 레벨에서 ID를 생성할 수 있고, 연속적인 시퀀스가 아니기 때문에 노출 시에도 공격 취약성이 덜할 것 같습니다.
3. `@GeneratedValue(strategy = GenerationType.AUTO)` 와 같이 사용하며 생성 전략에는 `AUTO`, `IDENTITY`, `SEQUENCE`, `TABLE` 이 있습니다.

soominsohn commented on src/main/java/roomescape/business/model/entity/Reservation.java:
```diff
@@ -0,0 +1,50 @@
+
+    @EmbeddedId
+    private final Id id;
+    @ManyToOne
+    private User user;
+    @Embedded
+    private ReservationDate date;
+    @ManyToOne
+    private ReservationTime time;
+    @ManyToOne
+    private Theme theme;
```
`@ManyToOne`을 사용하셨군요!
JPA에는 다양한 연관관계 매핑이 있는데, 그 중 `@ManyToOne`을 선택하신 이유가 궁금합니다.
또한,다른 객체들에서 Reservation을 가지고 있지 않고, Reservation이 연관된 객체들을 직접 가지고 있도록 설계하신 이유도 궁금해요. 😄

jye326 replied:
다른 객체들이 Reservation을 컬렉션 형태로 들고 있는 경우 각각에 대한 불필요한 컬렉션 관리 부담이 생깁니다. 또한 컬렉션에 OneToMany를 통한 양방향 참조가 필요하게 되는데 이는 동기화도 번거롭고, 메모리 부담도 있습니다. 이러한 이유로 일반적으로 단방향 ManyToOne을 사용한다고 생각하며, One이 Many 전체를 반드시 관리해야 하는 상황이 아니기에 OneToMany 형태를 지양했습니다.

soominsohn commented on src/main/java/roomescape/business/model/entity/User.java:
```diff
@@ -0,0 +1,55 @@
+@EqualsAndHashCode(of = "id")
+@Getter
+@Entity
+@Table(name = "users")
```
어떤 오류가 생겨 User객체는 `@Table` 어노테이션을 사용해 users 테이블과 매핑해주었나요?
다른 객체들도 `@Table` 어노테이션을 통해 테이블명을 명시적으로 표시해주면 더 안전해질 것 같아요.

jye326 replied:
애플리케이션에서 사용하는 Entity명을 그대로 Table명으로 활용하고 싶었으나 user 라는 네이밍은 이미 SQL에서 예약어로 지정되어 있어서 오류가 발생했습니다. 다른 엔티티 객체들에도 명시적으로 테이블 명을 표기했습니다!

soominsohn commented on src/main/resources/application.yml:
JPA 관련된 설정값도 application.yml 파일에서 에서 관리할 수 있는데요
어떤 설정값이 있고, 어떤 동작을하는지 학습해보면 좋을 것 같아요~

jye326 replied:
관련한 내용 찾아서 정리해보겠습니다!

soominsohn commented on src/main/java/roomescape/business/model/entity/User.java:
```diff
@@ -0,0 +1,55 @@
+    @EmbeddedId
+    private final Id id;
+    private UserRole userRole;
+    @Embedded
```
`@Embedded`, `@Embeddable` 는 어떤 역할을 하는 어노테이션일까요?

jye326 replied:
두 어노테이션은 VO를 필드로 가지는 엔티티를 매핑할때 사용합니다. `@Embedded`는 엔티티의 필드 위에 붙여 해당 필드가` @Embeddable`임을 명시합니다. VO의 필드값들은 엔티티의 칼럼으로 저장됩니다. `@Embeddable` 은 VO 클래스에 붙입니다

soominsohn replied:
좋습니다! VO를 칼럼으로 가진 엔티티를 저장할 때`AttributeConverter`라는 것도 사용하니 알아두시면 좋을 것 같아요~

soominsohn commented on src/main/java/roomescape/business/model/entity/User.java:
```diff
@@ -0,0 +1,55 @@
+
+    @EmbeddedId
+    private final Id id;
+    private UserRole userRole;
```
UserRole enum은 DB에 어떤 값으로 저장될까요?
저장되는 값이 의도하신 값일까요? 그렇다면 이유도 궁금합니다. 😄

jye326 replied:
확인해보니 기본 값인 `ORDINAL` 전략에 따라 0, 1로 저장되고 있었습니다. 별도로 의도한 부분은 없었습니다만, `@Enumerated(EnumType.STRING)`을 사용해 문자열로 저장하는 것이 데이터베이스 가독성을 높일 것 같아 변경하였습니다!

soominsohn commented on src/main/java/roomescape/infrastructure/JpaReservationDao.java:
```diff
@@ -0,0 +1,38 @@
+import java.time.LocalTime;
+import java.util.List;
+
+public interface JpaReservationDao extends JpaRepository<Reservation, Id> {
```
Controller-Service-Repository로 계층을 나누고,이번 미션에서 JDBC에서 JPA를 사용하게 리팩토링해주셨는데요,
계층을 분리해둔 덕분에 리팩토링 과정에서 얻은 장점이 있나요?
또, Jdbc → JPA로 변경하면서 어떤 장점과 단점을 느끼셨을까요?

jye326 replied:
계층 분리 덕에 다른 계층에서의 변경없이 리포지토리 구현체 만을 교체할 수 있었습니다. 또한 JPA를 사용하며 리포지토리 메서드 네이밍 기반 자동 생성 기능, 엔티티 매핑 기능 등의 편리함을 체감했습니다. Jdbc에 비한 단점은 아직 모르겠습니다. Jpa에 대해 공부해야 한다는 것이 단점일까요? 😏

soominsohn replied:
`Jpa에 대해 공부해야 한다는 것이 단점일까요? 😏` 공감이되네요 ㅎㅎㅎㅎ. 👍

soominsohn reviewed (changes requested):
안녕하세요 띠용~ 
만나서 반가워요. 이번 미션을 함께 할 앤지라고 합니다.
이번 미션 잘 구현해주셨네요. 코멘트 몇 개 남겼으니 확인 부탁드려요~

jye326 committed:
refactor(repository) : 패키지 분리

- JDBC, JPA 리포지토리 패키지 분리

jye326 committed:
fix(entity) : 기본 생성자 접근 지정 변경 -> protected

jye326 committed:
feat(entity) : 테이블 명 명시

jye326 committed:
feat(entity) : UserRole 저장 형식 변경 (ORDINAL-> STRING)

jye326 reviewed (commented):
빠르고 꼼꼼한 리뷰 감사합니다! 좋은 하루 되세요!!

soominsohn reviewed (changes requested):
띠용 안녕하세요~
리뷰 반영을 빠르게 해주셨네요. 👍
다만 이번 미션의 요구사항 중에 JDBC 의존성을 JPA 의존성으로 대체하는 항목이 포함되어 있어서,
해당 부분까지 수정해주시면 바로 다음 미션으로 넘어가도 좋을 것 같아요! 👍

jye326 committed:
refactor: Jdbc 레포지토리 코드 제거 및 Jpa 레포지토리로 통합

- Jdbc 기반의 Reservation, ReservationTime, Theme, User 레포지토리 코드를 삭제하고 기존 Jpa 기반 레포지토리로 통합.
- 관련 테스트 코드 및 인프라 패키지 구조에서 불필요한 "Jpa" 네이밍 제거로 간결화 작업 수행.

jye326 committed:
refactor: Jdbc 레포지토리 코드 제거 및 Jpa 레포지토리로 통합

- Jdbc 기반의 Reservation, ReservationTime, Theme, User 레포지토리 코드를 삭제하고 기존 Jpa 기반 레포지토리로 통합.
- 관련 테스트 코드 및 인프라 패키지 구조에서 불필요한 "Jpa" 네이밍 제거로 간결화 작업 수행.

soominsohn reviewed (approved):
띠용 안녕하세요~
굉장히 빠르게 반영해주셨네요 👍 
그럼 이제 다음 단계로 넘어가볼까요?
기대하고 있겠습니다. 파이팅하세요~
